src/expr.rs

```diff
@@ -423,22 +426,101 @@ fn rewrite_if_else(context: &RewriteContext,
+                rewrite_if_else(context,
+                                cond,
+                                if_block,
+                                else_block.as_ref().map(|e| &**e),
+                                Some(pat),
+                                width,
+                                offset,
+                                false)
+            }
+            ast::Expr_::ExprIf(ref cond, ref if_block, ref else_block) => {
+                rewrite_if_else(context,
+                                cond,
+                                if_block,
+                                else_block.as_ref().map(|e| &**e),
+                                None,
+                                width,
+                                offset,
+                                false)
+            }
+            _ => else_block.rewrite(context, width, offset),
+        };
 
         result.push_str(" else ");
-        result.push_str(&else_block_string);
+        result.push_str(&&try_opt!(rewrite));
     }
 
     Some(result)
 }
 
+fn single_line_if_else(context: &RewriteContext,
+                       pat_expr_str: &str,
+                       if_node: &ast::Block,
+                       else_block_opt: Option<&ast::Expr>,
+                       width: usize)
+                       -> Option<String> {
+    let else_block = try_opt!(else_block_opt);
+    let fixed_cost = "if  {  } else {  }".len();
+
+    if let ast::ExprBlock(ref else_node) = else_block.node {
+        if !is_simple_block(if_node, context.codemap) ||
+           !is_simple_block(else_node, context.codemap) || pat_expr_str.contains('\n') {
+            return None;
+        }
+
+        let new_width = try_opt!(width.checked_sub(pat_expr_str.len() + fixed_cost));
+        let if_expr = if_node.expr.as_ref().unwrap();
+        let if_str = try_opt!(if_expr.rewrite(context, new_width, 0));
+
+        let new_width = try_opt!(new_width.checked_sub(if_str.len()));
+        let else_expr = else_node.expr.as_ref().unwrap();
+        let else_str = try_opt!(else_expr.rewrite(context, new_width, 0));
+
+        // FIXME: this check shouldn't be necessary. Rewrites should either fail
+        // or wrap to a newline when the object does not fit the width.
+        let fits_line = fixed_cost + pat_expr_str.len() + if_str.len() + else_str.len() <= width;
+
+        if fits_line && !if_str.contains('\n') && !else_str.contains('\n') {
+            return Some(format!("if {} {{ {} }} else {{ {} }}", pat_expr_str, if_str, else_str));
+        }
+    }
+
+    None
+}
+
+// Checks that a block contains no statements, an expression and no comments.
+fn is_simple_block(block: &ast::Block, codemap: &CodeMap) -> bool {
+    if !block.stmts.is_empty() || block.expr.is_none() {
+        return false;
+    }
+
+    let snippet = codemap.span_to_snippet(block.span).unwrap();
+
+    !snippet.contains("//") && !snippet.contains("/*")
```

Comment:
Hmm, I am not sure I understand. It will fail when `'//` is used as either a label or a lifetime, you mean? Would that be possible?